File: ios/SharedStorage.m
#import <Foundation/Foundation.h>
#import <React/RCTBridgeModule.h>

@interface SharedStorage : NSObject <RCTBridgeModule>
@end

@implementation SharedStorage

RCT_EXPORT_MODULE();

- (NSString *)suiteName {
  return @"group.com.yusuke.dmpvp";
}

RCT_REMAP_METHOD(setString,
                 setString:(NSString *)key
                 value:(NSString *)value
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
  NSUserDefaults *defaults = [[NSUserDefaults alloc] initWithSuiteName:[self suiteName]];
  if (!defaults) {
    reject(@"no_defaults", [NSString stringWithFormat:@"UserDefaults suite not found: %@", [self suiteName]], nil);
    return;
  }

  [defaults setObject:value forKey:key];
  [defaults synchronize];

  // ✅ WidgetCenterは使わない（リンク問題回避）
  resolve(@(YES));
}

RCT_REMAP_METHOD(getString,
                 getString:(NSString *)key
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
  NSUserDefaults *defaults = [[NSUserDefaults alloc] initWithSuiteName:[self suiteName]];
  if (!defaults) {
    reject(@"no_defaults", [NSString stringWithFormat:@"UserDefaults suite not found: %@", [self suiteName]], nil);
    return;
  }

  NSString *v = [defaults stringForKey:key];
  resolve(v ?: @"");
}

@end

File: ios/SharedStorage.swift
import Foundation
import WidgetKit

@objc(SharedStorage)
class SharedStorage: NSObject {

  private let suiteName = "group.com.yusuke.dmpvp"

  @objc
  func setString(_ key: String,
                 value: String,
                 resolver resolve: RCTPromiseResolveBlock,
                 rejecter reject: RCTPromiseRejectBlock) {

    guard let defaults = UserDefaults(suiteName: suiteName) else {
      reject("no_defaults", "UserDefaults suite not found: \(suiteName)", nil)
      return
    }

    defaults.set(value, forKey: key)
    defaults.synchronize()

    // ✅ Widgetを更新
    WidgetCenter.shared.reloadAllTimelines()

    resolve(true)
  }

  @objc
  func getString(_ key: String,
                 resolver resolve: RCTPromiseResolveBlock,
                 rejecter reject: RCTPromiseRejectBlock) {

    guard let defaults = UserDefaults(suiteName: suiteName) else {
      reject("no_defaults", "UserDefaults suite not found: \(suiteName)", nil)
      return
    }

    resolve(defaults.string(forKey: key) ?? "")
  }

  @objc static func requiresMainQueueSetup() -> Bool { false }
}
